crates/prover/src/components/memory.rs

```diff
@@ -32,88 +34,142 @@ pub struct InteractionClaim {
 }
 
 pub struct LookupData {
-    pub memory: Vec<[PackedM31; 3]>,
+    pub initial_memory: Vec<[PackedM31; N_M31_IN_MEMORY_ENTRY]>,
+    pub final_memory: Vec<[PackedM31; N_M31_IN_MEMORY_ENTRY]>,
 }
 
 impl Claim {
+    pub fn new(memory_boundaries: MemoryBoundaries) -> Self {
+        let column_length = std::cmp::max(
+            std::cmp::max(
+                memory_boundaries.initial_memory.len(),
+                memory_boundaries.final_memory.len(),
+            )
+            .next_power_of_two(),
+            N_LANES,
+        );
+        let log_size = column_length.ilog2();
+
+        Self {
+            log_size,
+            inputs: memory_boundaries,
+        }
+    }
+
     pub fn log_sizes(&self) -> TreeVec<Vec<u32>> {
-        let trace_log_sizes = vec![self.log_size; 3];
-        let interaction_log_sizes = vec![self.log_size; SECURE_EXTENSION_DEGREE];
+        let trace_log_sizes = vec![self.log_size; 2 * N_M31_IN_MEMORY_ENTRY];
+        let interaction_log_sizes = vec![self.log_size; SECURE_EXTENSION_DEGREE * 2];
         TreeVec::new(vec![vec![], trace_log_sizes, interaction_log_sizes])
     }
 
     pub fn mix_into(&self, channel: &mut impl Channel) {
         channel.mix_u64(self.log_size as u64);
     }
 
-    #[allow(non_snake_case)]
     pub fn write_trace<MC: MerkleChannel>(
         &mut self,
-        lookup_data: &Vec<[PackedM31; 2]>,
     ) -> (
-        [CircleEvaluation<SimdBackend, M31, BitReversedOrder>; 3],
+        Vec<CircleEvaluation<SimdBackend, M31, BitReversedOrder>>,
         LookupData,
     )
     where
         SimdBackend: BackendForChannel<MC>,
     {
-        let mut counts: HashMap<(u32, u32), u32> = HashMap::new();
-        for entry in lookup_data {
-            let arr0 = entry[0].to_array();
-            let arr1 = entry[1].to_array();
-            for (val0, val1) in arr0.iter().zip(arr1.iter()) {
-                let key = (val0.0, val1.0);
-                *counts.entry(key).or_insert(0) += 1;
+        // Pack memory entries from the prover input
+        let initial_memory: Vec<[PackedM31; N_M31_IN_MEMORY_ENTRY]> = self
```

Comment:
pardon my ignorance but what is inside the trace of the memory component? it looks like we only 'log' the memory at the initial step and the final step?

are we not supposed to log every single memory access?

Reply:
opcode components "clear" and "write" memory values, the memory component is only useful for boundary conditions indeed

Reply:
and checking that "squashing" iterative prev_value -> value is consistent?